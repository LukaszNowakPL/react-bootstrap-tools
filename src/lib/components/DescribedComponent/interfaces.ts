export interface DescribedComponentProps {
  children: any;
  description?: string;
  className?: string;
  placement?: PlacementValues;
  delay?: number;
  defaultShow?: boolean;
  trigger?: TriggerValues;
}

export interface ConfigInterface {
  defaultTooltipClassName: string;
  defaultTooltipId: string;
  defaultTooltipPlacement: PlacementValues;
}

export interface TooltipProps {
  id: string;
  className: string;
}

export interface OverlayTriggerProps {
  defaultShow?: boolean;
  delay?: number;
  overlay: any;
  trigger?: TriggerValues;
  placement?: PlacementValues;
}

export type TriggerValues =
  | "hover"
  | "click"
  | "focus"
  | Array<"hover" | "click" | "focus">;

export type PlacementValues =
  | "auto-start"
  | "auto"
  | "auto-end"
  | "top-start"
  | "top"
  | "top-end"
  | "right-start"
  | "right"
  | "right-end"
  | "bottom-end"
  | "bottom"
  | "bottom-start"
  | "left-end"
  | "left"
  | "left-start";
