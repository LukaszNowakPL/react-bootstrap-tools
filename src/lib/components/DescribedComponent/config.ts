import {ConfigInterface} from "./interfaces";

const config:ConfigInterface = {
  defaultTooltipClassName: "Tooltip",
  defaultTooltipId: "tooltip",
  defaultTooltipPlacement: "top"
};

export default config;
