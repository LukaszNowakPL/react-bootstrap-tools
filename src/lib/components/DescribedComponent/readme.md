# DescribedComponent

Wrapper for describing given html portion using Bootstrap's `Tooltip` component.

![DescribedComponent example](../../../../readme_img/DescribedComponentExample.png)

## Example of usage

```
import React from 'react';
import { DescribedComponent } from 'react-bootstrap-tools';

const SimplestExample = () => (
    <DescribedComponent description="description text">
        <p>html element</p>
    </DescribedComponent>
);

export default SimplestExample;
```

## Available optional props

- string `description` : Description text to put inside `Tooltip` component
- string `className` : Additional class name to custom your styling of a `Tooltip` component.
  Styling example for `NewStyle`:<pre>
  .NewStyle > .tooltip-inner {
  background-color: yellow;
  color: black;
  }
  .NewStyle.bs-tooltip-top .arrow::before {
  border-top-color: yellow !important;
  }
  .NewStyle.bs-tooltip-right .arrow::before {
  border-right-color: yellow !important;
  }
  .NewStyle.bs-tooltip-bottom .arrow::before {
  border-bottom-color: yellow !important;
  }
  .NewStyle.bs-tooltip-left .arrow::before {
  border-left-color: yellow !important;
  }</pre>
- string `placement` : Position of a `Tooltip` over component it describes.<br />
  Available options:
  - auto-start
  - auto
  - auto-end
  - top-start
  - top
  - top-end
  - right-start
  - right
  - right-end
  - bottom-end
  - bottom
  - bottom-start
  - left-end
  - left
  - left-start
- number `delay` : Number of milliseconds the `Tooltip` component will appear and hide after describing component hover.
- boolean `defaultShow` : Set `true` will initialise component with visible description
- string `trigger` : Action necessary to take over component to initialise visibility of description.<br />
  Available options:
  - hover
  - click
  - focus
  - Array<hover, click, focus>
