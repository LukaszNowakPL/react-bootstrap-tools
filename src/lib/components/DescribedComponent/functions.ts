import {
  DescribedComponentProps,
  OverlayTriggerProps,
  TooltipProps
} from "./interfaces";
import config from "./config";

/**
 * Returns props used by OverlayTrigger component
 * @param props
 */
export const getOverlayTriggerProps = (props: DescribedComponentProps) => {
  const overlayTriggerProps: OverlayTriggerProps = {
    overlay: null
  };

  if (props.defaultShow) overlayTriggerProps.defaultShow = props.defaultShow;
  if (props.delay) overlayTriggerProps.delay = props.delay;
  if (props.trigger) overlayTriggerProps.trigger = props.trigger;
  if (props.placement) overlayTriggerProps.placement = props.placement;

  return overlayTriggerProps;
};

/**
 * Returns props used by Tooltip component
 * @param props
 */
export const getTooltipProps = (props: DescribedComponentProps) => {
  const tooltipProps: TooltipProps = {
    id: config.defaultTooltipId,
    className: getTooltipClassName(
      config.defaultTooltipClassName,
      props.className
    )
  };

  return tooltipProps;
};

/**
 * Returns class name of a Tooltip component
 * @param defaultClassName
 * @param className - additional class name
 */
const getTooltipClassName = (
  defaultClassName: string,
  className: string | undefined
): string => {
  if (className) return defaultClassName + " " + className;

  return defaultClassName;
};
