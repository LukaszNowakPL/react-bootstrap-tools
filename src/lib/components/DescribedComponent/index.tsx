import * as React from "react";
import {
  DescribedComponentProps,
  OverlayTriggerProps,
  TooltipProps
} from "./interfaces";
import { getOverlayTriggerProps, getTooltipProps } from "./functions";
import { OverlayTrigger, Tooltip } from "react-bootstrap";

const DescribedComponent = (props: DescribedComponentProps) => {
  const { children, description } = props;

  if (description) {
    const tooltipProps: TooltipProps = getTooltipProps(props);

    const overlayTriggerProps: OverlayTriggerProps = getOverlayTriggerProps(
      props
    );

    const overlay = <Tooltip {...tooltipProps}>{description}</Tooltip>;

    return (
      <OverlayTrigger {...overlayTriggerProps} overlay={overlay}>
        {children}
      </OverlayTrigger>
    );
  }
  return <span>{props.children}</span>;
};

export default DescribedComponent;
