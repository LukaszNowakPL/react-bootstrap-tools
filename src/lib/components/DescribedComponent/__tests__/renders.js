import React from "react";
import { shallow } from "enzyme";
import DescribedComponent from "../index";

describe("<DescribedComponent />", () => {
  const customComponentProps = {
    description: "test description"
  };

  it("renders", () => {
    const component = shallow(
      <DescribedComponent {...customComponentProps}>
        <p>child</p>
      </DescribedComponent>
    );
    expect(component).toMatchSnapshot();
  });

  it("renders with default show", () => {
    const component = shallow(
      <DescribedComponent {...customComponentProps} defaultShow={true}>
        <p>child</p>
      </DescribedComponent>
    );
    expect(component).toMatchSnapshot();
  });

  it("renders with custom show/hide delay", () => {
    const component = shallow(
      <DescribedComponent {...customComponentProps} delay={3000}>
        <p>child</p>
      </DescribedComponent>
    );
    expect(component).toMatchSnapshot();
  });

  it("renders with custom trigger", () => {
    const component = shallow(
      <DescribedComponent {...customComponentProps} trigger="hover">
        <p>child</p>
      </DescribedComponent>
    );
    expect(component).toMatchSnapshot();
  });

  it("renders with custom position", () => {
    const component = shallow(
      <DescribedComponent {...customComponentProps} placement="top">
        <p>child</p>
      </DescribedComponent>
    );
    expect(component).toMatchSnapshot();
  });

  it("renders with custom style", () => {
    const component = shallow(
      <DescribedComponent {...customComponentProps} clasName="test className">
        <p>child</p>
      </DescribedComponent>
    );
    expect(component).toMatchSnapshot();
  });

  it("renders with no description", () => {
    const component = shallow(
      <DescribedComponent>
        <p>child</p>
      </DescribedComponent>
    );
    expect(component).toMatchSnapshot();
  });
});
