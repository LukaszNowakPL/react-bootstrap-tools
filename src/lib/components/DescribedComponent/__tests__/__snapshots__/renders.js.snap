// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DescribedComponent /> renders 1`] = `
<OverlayTrigger
  defaultOverlayShown={false}
  overlay={
    <ForwardRef(Bootstrap(Tooltip))
      className="Tooltip"
      id="tooltip"
    >
      test description
    </ForwardRef(Bootstrap(Tooltip))>
  }
  trigger={
    Array [
      "hover",
      "focus",
    ]
  }
>
  <p>
    child
  </p>
</OverlayTrigger>
`;

exports[`<DescribedComponent /> renders with custom position 1`] = `
<OverlayTrigger
  defaultOverlayShown={false}
  overlay={
    <ForwardRef(Bootstrap(Tooltip))
      className="Tooltip"
      id="tooltip"
    >
      test description
    </ForwardRef(Bootstrap(Tooltip))>
  }
  placement="top"
  trigger={
    Array [
      "hover",
      "focus",
    ]
  }
>
  <p>
    child
  </p>
</OverlayTrigger>
`;

exports[`<DescribedComponent /> renders with custom show/hide delay 1`] = `
<OverlayTrigger
  defaultOverlayShown={false}
  delay={3000}
  overlay={
    <ForwardRef(Bootstrap(Tooltip))
      className="Tooltip"
      id="tooltip"
    >
      test description
    </ForwardRef(Bootstrap(Tooltip))>
  }
  trigger={
    Array [
      "hover",
      "focus",
    ]
  }
>
  <p>
    child
  </p>
</OverlayTrigger>
`;

exports[`<DescribedComponent /> renders with custom style 1`] = `
<OverlayTrigger
  defaultOverlayShown={false}
  overlay={
    <ForwardRef(Bootstrap(Tooltip))
      className="Tooltip"
      id="tooltip"
    >
      test description
    </ForwardRef(Bootstrap(Tooltip))>
  }
  trigger={
    Array [
      "hover",
      "focus",
    ]
  }
>
  <p>
    child
  </p>
</OverlayTrigger>
`;

exports[`<DescribedComponent /> renders with custom trigger 1`] = `
<OverlayTrigger
  defaultOverlayShown={false}
  overlay={
    <ForwardRef(Bootstrap(Tooltip))
      className="Tooltip"
      id="tooltip"
    >
      test description
    </ForwardRef(Bootstrap(Tooltip))>
  }
  trigger="hover"
>
  <p>
    child
  </p>
</OverlayTrigger>
`;

exports[`<DescribedComponent /> renders with default show 1`] = `
<OverlayTrigger
  defaultOverlayShown={false}
  defaultShow={true}
  overlay={
    <ForwardRef(Bootstrap(Tooltip))
      className="Tooltip"
      id="tooltip"
    >
      test description
    </ForwardRef(Bootstrap(Tooltip))>
  }
  trigger={
    Array [
      "hover",
      "focus",
    ]
  }
>
  <p>
    child
  </p>
</OverlayTrigger>
`;

exports[`<DescribedComponent /> renders with no description 1`] = `
<span>
  <p>
    child
  </p>
</span>
`;
